packages/astro/src/actions/runtime/virtual/shared.ts

```diff
@@ -42,7 +42,7 @@ const statusToCodeMap: Record<number, ActionErrorCode> = Object.entries(codeToSt
 
 export type ErrorInferenceObject = Record<string, any>;
 
-export class ActionError<T extends ErrorInferenceObject = ErrorInferenceObject> extends Error {
+export class ActionError extends Error {
```

Comment:
This `T` was unused 🔪 . And, as a result, `SafeResult`'s `TInput` later on becomes unused as well. 🔪

Reply:
lgtm, @bholmesdev can you confirm?

Reply:
I've documented the reason for this generic type [here](https://github.com/withastro/astro/blob/main/packages/astro/src/actions/runtime/utils.ts#L37-L46). The generic is used by `ActionInputError` (a descendant of `ActionError`). However, I've found that if you remove this generic from `ActionError`, all type inference breaks for our `isInputError()` utility. If you have a suggestion to avoid this, please let me know @JoshuaKGoldberg. We now have a [type test](https://github.com/withastro/astro/blob/main/packages/astro/test/types/is-input-error.ts) to assert whether removing this generic causes any issue

Reply:
Aha! Thanks, that's helpful.

The root problem here is that your `isInputError` type is not type-safe. It violates the ["Golden Rule of Generics"](https://effectivetypescript.com/2020/08/12/generics-golden-rule). Nothing in the body of `isInputError` actually uses the type parameter `T`. You could call it like...

```ts
const example = new ActionError({ code: "BAD_REQUEST" });

if (isInputError<{ gotcha: "whoops" }>(example) {
	example.fields.gotcha;
	//             ^? string[] | undefined

}
```

...and TypeScript wouldn't know to complain.

> Aside: which we recently codified that golden rule in typescript-eslint as [`@typescript-eslint/no-unnecessary-type-parameters`](https://typescript-eslint.io/rules/no-unnecessary-type-parameters)... but the lint rule doesn't (yet?) have the ability to check type parameters passed as type arguments - see its [Limitations](https://typescript-eslint.io/rules/no-unnecessary-type-parameters/#limitations)). Pity. This wouldn't be caught by linting right now.

I don't have enough context to refactor things to not violate the golden rule. Sometimes the right approach actually is to just `eslint-disable-next-line` the unused thing and/or only-used-once type parameter. So I'll revert these removals for now with a comment.